silverback-integration-kafka/src/Silverback.Integration.Kafka/Messaging/Broker/KafkaProducer.cs

```diff
@@ -0,0 +1,65 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Text;
+using Confluent.Kafka;
+using Confluent.Kafka.Serialization;
+using Silverback.Messaging.Messages;
+
+namespace Silverback.Messaging.Broker
+{
+    /// <summary>
+    /// A Apache Kafka bases <see cref="IProducer" />
+    /// </summary>
+    /// <seealso cref="Silverback.Messaging.Broker.Producer" />
+    /// <seealso cref="Silverback.Messaging.Broker.IProducer" />
+    public class KafkaProducer : Producer
+    {
+        private readonly Producer<byte[], byte[]> _producer;
+        private readonly string _topic;
+        private readonly IDeliveryHandler<byte[], byte[]> _deliveryHandler;
+
+
+        /// <summary>
+        /// Kafka producer constructor
+        /// </summary>
+        /// <param name="endpoint"></param>
+        /// <param name="config"></param>
+        /// <param name="handler"></param>
+        public KafkaProducer(IEndpoint endpoint, Dictionary<string, object> config, IDeliveryHandler<byte[], byte[]> handler = null)
```

Comment:
The delivery handler shouldn't be provided via factory? I don't think it's handy to have to reuse the same instance for all messages, if you have dependencies like the DbContext in the IDeliveryHandler implementation. Maybe it doesn't make sense to have such dependency, let's discuss. --> Maybe it's the idea behind the default implementation of DeliveryHandler, it's just not clear (yet).

Plus, just an idea, shouldn't we provide an abstraction over the IDeliveryHandler, to have a an HandleDeliveryReport method that receives our deserialized messages instead of the byte[]?

Reply:
IDeliveryHandler is an Interface of Confluent.Kafka. This is a callback interface used by ProducerAsync method. Our DeliveryHandler implementation have a constructor parameter Func<IDeliveryResultHandler> _handler.

```
public void HandleDeliveryReport(Message<byte[], byte[]> message)
        {
            using (var processor = _handler.Invoke())
            {
                processor.ProcessDelivery(message);
            }
        }  
```